nautobot/ipam/utils/migrations.py

```diff
@@ -0,0 +1,569 @@
+    return [f"{network}/{prefix_length}" for network, prefix_length in dupes]
+
+
+def generate_parent_prefix(apps, address):
+    """For a given `address`, generate a containing parent network address."""
+    cidr = validate_cidr(apps, address)
+    return cidr.cidr
+
+
+def get_closest_parent(apps, obj, qs):
+    """
+    This is forklifted from `Prefix.objects.get_closest_parent()` so that it can safely be used in
+    migrations.
+
+    Return the closest matching parent Prefix for a `cidr` even if it doesn't exist in the database.
+
+    Args:
+        obj: Prefix/IPAddress instance
+        namespace (Namespace): Namespace instance
+        max_prefix_length (int): Maximum prefix length depth for closest parent lookup
+        tenant (Tenant): Tenant instance
+    """
+    # Validate that it's a real CIDR
+    cidr = validate_cidr(apps, obj)
+    broadcast = str(cidr.broadcast or cidr.ip)
+
+    Prefix = apps.get_model("ipam", "Prefix")
+    IPAddress = apps.get_model("ipam", "IPAddress")
+
+    # Prepare the queryset filter
+    lookup_kwargs = {
+        "ip_version": cidr.version,
+        "network__lte": cidr.network,
+        "broadcast__gte": broadcast,
+    }
+
+    if isinstance(obj, Prefix):
+        lookup_kwargs["prefix_length__lt"] = cidr.prefixlen
+        qs = qs.exclude(id=obj.id)
+    else:
+        lookup_kwargs["prefix_length__lte"] = cidr.prefixlen
+
+    # Search for possible ancestors by network/prefix, returning them in
+    # reverse prefix length order, so that we can choose the first one.
+    possible_ancestors = (
+        qs.filter(**lookup_kwargs)
+        .annotate(
+            custom_sort_order=models.Case(
+                models.When(tenant=obj.tenant, vrf=obj.vrf, then=models.Value(1)),
+                models.When(tenant__isnull=True, vrf=obj.vrf, then=models.Value(2)),
+                models.When(tenant=obj.tenant, vrf__isnull=True, then=models.Value(3)),
+                models.When(vrf=obj.vrf, then=models.Value(4)),
+                models.When(tenant__isnull=True, vrf__isnull=True, then=models.Value(5)),
+                models.When(vrf__isnull=True, then=models.Value(6)),
+                default=models.Value(7),
+            )
+        )
+        .order_by("-prefix_length", "custom_sort_order")
+    )
+
+    if isinstance(obj, IPAddress):
+        # IP should not fall back to less specific prefixes
+        if not possible_ancestors.exists():
+            return qs.none()
+        prefix_length = possible_ancestors.first().prefix_length
+        return possible_ancestors.filter(prefix_length=prefix_length)
+
+    # If we've got any matches, the first one is our closest parent.
+    try:
+        return possible_ancestors[0]
+    except IndexError:
+        raise Prefix.DoesNotExist(f"Could not determine parent Prefix for {cidr}")
+
+
+def get_next_cleanup_namespace(apps, vrf):
+    """
+    Try to get the next available Cleanup Namespace based on `vrf`.
```

Comment:
If I'm reading right, this function is only used for VRFs that are currently part of the `Global` namespace? Seems like it would fall down if this is not the case - would be good to document that assumption, maybe even add an assert or two if so.